src/lib/FreeC/Backend/Agda/Converter/Type.hs

```diff
@@ -66,10 +98,47 @@ dagger = fmap free . star
 --   > (τ₁ → τ₂)* = τ₁' → τ₂'
 --   > C* = Ĉ Shape Position
 --   > α* = α̂
-star :: IR.Type -> Converter Agda.Expr
-star (IR.TypeVar s name) = Agda.Ident
-  <$> lookupAgdaIdentOrFail s IR.TypeScope (IR.UnQual (IR.Ident name))
-star (IR.TypeCon s name) =
+convertSizedType' :: Maybe Agda.Expr -> IR.Type -> Converter Agda.Expr
+convertSizedType' (Just i) t = Agda.app <$> convertType' t <*> pure i
+convertSizedType' _ (IR.TypeVar s name) = Agda.Ident
+  <$> lookupAgdaIdentOrFail s IR.TypeScope (IR.UnQual $ IR.Ident name)
+convertSizedType' _ (IR.TypeCon s name) =
   applyFreeArgs <$> lookupAgdaIdentOrFail s IR.TypeScope name
-star (IR.TypeApp  _ l r) = Agda.app <$> star l <*> star r
-star (IR.FuncType _ l r) = Agda.fun <$> dagger l <*> dagger r
+convertSizedType' _ (IR.TypeApp _ l r) =
+  Agda.app <$> convertType' l <*> convertType' r
+convertSizedType' _ (IR.FuncType _ l r) =
+  Agda.fun <$> convertType l <*> convertType r
```

Comment:
If I understand this code correctly a size annotation `i` is added to `τ` by generating `τ* i`. However in `convertTypeDecl`, the size is a hidden argument. Thus, I think `τ* {i}` should be generated. I'm also not sure whether it is always correct to put the size annotation at the outermost position. For example if `τ` is a type variable, it cannot be applied to a size. It may also be necessary to add size annotations to the arguments of type constructor applications and to function types recursively.

Reply:
The `Maybe Agda.Expr` is the already hidden type variable. [`convertRecFuncType`](https://github.com/FreeProving/free-compiler/blob/84cf653e30f6a080ab849721c3acca597b36cb04/src/lib/FreeC/Backend/Agda/Converter/Type.hs#L45) and [`convertRecConType`](https://github.com/FreeProving/free-compiler/blob/84cf653e30f6a080ab849721c3acca597b36cb04/src/lib/FreeC/Backend/Agda/Converter/Type.hs#L68) call the transformations with the already hidden argument. Since all type annotations are hidden it seems reasonable to move the `Agda.hiddenArg_` call to `convertSizedType'`.
`convertSizedType` and `convertSizedType'` are always called from inside this module. I therefore assumed that `τ` is a type constructor if a size variable is passed. Applying the size variable to other nodes in the AST is of course possible (the first version of my code took a variable of type `Maybe (Agda.QName, Agda.Expr)` and searched for type constructors with the given name to apply the variable to; describing where to put an annotation shouldn't be a problem).
There are of course cases where the decreasing argument is inside another type. The following Agda code does type check and pass termination checks. 
```agda
data Pairₜ (S : Set) (P : S → Set) (A B : Set) : Set where
  pair : Free S P A → Free S P B → Pairₜ S P A B

pattern Pair x y = pure (pair x y)

weirdMap : ∀ {S P A B C i} → Free S P (Free S P A → Free S P B) 
                           → Free S P (Pairₜ S P (List S P A {i}) C) 
                           → Free S P (Pairₜ S P (List S P B) C)
weirdMap mf acs = acs >>= λ where
  (pair as mc) → as >>= λ where
    nil           → Pair [] mc
    (cons mx mxs) → weirdMap mf (Pair mxs mc) >>= λ where
      (pair mbs mc) → mf >>= λ where
        f → Pair (f mx ∷ mbs) mc
```
The issue with examples like this one is of course finding the decreasing argument (because it is passed to a constructor before the recursive call) and then linking it back to a part of the type.
The problem I see with applying the same size annotations to more then type constructor is constraining the type of the function. Assume `f` take an argument of type `List A i × List B i` but decreases just on the first component of the pair. The recursive call would occur on a variable of type `List A j × List B i` with `j < i` which does not type check, because `i` and `j` are clearly not equal.